lib/matplotlib/layoutbox.py

```diff
@@ -0,0 +1,727 @@
+        #     sol.addEditVariable(self.left_margin, 'weak')
+        #     sol.suggestValue(self.left_margin, 0.0001)
+        c = (self.left_margin == self.left - self.parent.left)
+        self.solver.addConstraint(c | 'required')
+        c = (self.left_margin >= self.left_margin_min)
+        self.solver.addConstraint(c | 'strong')
+
+        # right
+        if not sol.hasEditVariable(self.right_margin_min):
+            sol.addEditVariable(self.right_margin_min, 'strong')
+            sol.suggestValue(self.right_margin_min, 0.0001)
+        # if not sol.hasEditVariable(self.right_margin):
+        #     sol.addEditVariable(self.right_margin, 'weak')
+        #     sol.suggestValue(self.right_margin, 0.0001)
+        c = (self.right_margin == self.parent.right - self.right)
+        self.solver.addConstraint(c | 'required')
+        c = (self.right_margin >= self.right_margin_min)
+        self.solver.addConstraint(c | 'required')
+        # bottom
+        if not sol.hasEditVariable(self.bottom_margin_min):
+            sol.addEditVariable(self.bottom_margin_min, 'strong')
+            sol.suggestValue(self.bottom_margin_min, 0.0001)
+        # if not sol.hasEditVariable(self.bottom_margin):
+        #     sol.addEditVariable(self.bottom_margin, 'weak')
+        #     sol.suggestValue(self.bottom_margin, 0.0001)
+        c = (self.bottom_margin == self.bottom - self.parent.bottom)
+        self.solver.addConstraint(c | 'required')
+        c = (self.bottom_margin >= self.bottom_margin_min)
+        self.solver.addConstraint(c | 'required')
+        # top
+        if not sol.hasEditVariable(self.top_margin_min):
+            sol.addEditVariable(self.top_margin_min, 'strong')
+            sol.suggestValue(self.top_margin_min, 0.0001)
+        # if not sol.hasEditVariable(self.top_margin):
+        #     sol.addEditVariable(self.top_margin, 'weak')
+        #     sol.suggestValue(self.top_margin, 0.0001)
+        c = (self.top_margin == self.parent.top - self.top)
+        self.solver.addConstraint(c | 'required')
+        c = (self.top_margin >= self.top_margin_min)
+        self.solver.addConstraint(c | 'required')
+
+    def add_child(self, child):
+        self.children += [child]
+
+    def remove_child(self, child):
+        try:
+            self.children.remove(child)
+        except ValueError:
+            print("Tried to remove child that doesn't belong to parent")
+
+    def add_constraints(self):
+        sol = self.solver
+        # never let width and height go negative.
+        for i in [self.min_width, self.min_height]:
+            sol.addEditVariable(i, 1e9)
+            sol.suggestValue(i, 0.0)
+        # define relation ships between things thing width and right and left
+        self.hard_constraints()
+        # self.soft_constraints()
+        if self.parent:
+            self.parent_constrain()
+        # sol.updateVariables()
+
+    def parent_constrain(self):
+        parent = self.parent
+        eps = 0.0000000
```

Comment:
hum?

Reply:
Well, you know, it wasn't zero for a while ;-)